spec/jmap/push.mdown

```diff
@@ -1,69 +1,94 @@
 
-### setPushCallback
+### setPushSubscription
 
-To set the web hook, make a call to *setPushCallback*. It takes the following argument:
+Each session may only have a single push subscription registered. The push subscription is tied to the access token used to create it. Should the access token expire or be revoked, the push subscription MUST be removed by the JMAP server. The client MUST re-register the push subscription after reauthenticating to resume callbacks.
 
-- **callback**: `String|null`
-  The (HTTPS) URL the JMAP server should POST events to. This will replace any previously set URL. Set to `null` to just remove any previously set callback URL.
+To set the push subscription, make a call to *setPushSubscription*. It takes the following argument:
 
-The response to *setPushCallback* is called *pushCallbackSet*. It has the following argument:
+- **pushSubscription**: `PushSubscription|null`
+  The PushSubscription object representing the endpoint the JMAP server will POST events to. This will replace any previously set subscription. Set to `null` to remove any previously registered subscription.
 
-- **callback**: `String|null`
-  Echoed back from the call.
+The response to *setPushSubscription* is called *pushSubscriptionSet*. It has no arguments.
 
-The following error may be returned instead of the *mailboxesSet* response:
+The following errors may be returned instead of the *pushSubscriptionSet* response:
 
 `invalidUrl`: Returned if the URL does not begin with `https://`, or is otherwise syntactically invalid or does not resolve.
 
 `forbidden`: Returned if the URL is valid, but for policy reasons the server is not willing to connect to it.
 
-### getPushCallback
+### getPushSubscription
+
+To check the currently set push subscription (if any), make a call to *getPushSubscription*. It does not take any arguments. The response is called *pushSubscription* and it has a single argument:
+
+- **pushSubscription**: `PushSubscription|null`
+  The PushSubscription object the JMAP server is currently posting push events to, or `null` if none.
+
+## Event Source
+
+Clients that can hold open TCP connections can connect directly to the JMAP server to receive push notifications via a `text/event-stream` resource, as described in <http://www.w3.org/TR/eventsource/>. This is a long running HTTP request down which the server can push data.
+
+When a change occurs in the data on the server, it pushes an event called **state** to any connected clients, with the *StateChange* object as the data.
+
+The server SHOULD also set a new `Last-Event-Id` that encodes the entire server state visible to the user. When a new connection is made to the event-source endpoint, the server can then work out whether the client has missed some changes which it should send immediately.
```

Comment:
I don't have enough context to understand how to use this Last-Event-Id.  Do you have a reference to other parts of the document, or can you expand the description. (Where does the server set this? Does the client set something?)

Reply:
The concept is a standard part of [Server Sent Events](https://html.spec.whatwg.org/multipage/server-sent-events.html), which is why I didn't describe it in too much detail, but I can add more.

Basically, the server can send an id with an event in the text/event-stream. If the client loses its connection, when it reconnects it sends the server the last id it saw as a Last-Event-ID header in the HTTP request. The server can then detect if the client missed any events while it was disconnected and send them immediately if so.